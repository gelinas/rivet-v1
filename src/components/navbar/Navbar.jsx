import React, { useState, useRef } from 'react';
import { Link } from 'react-router-dom';

//styles
import './navbar.scss'

export default function Navbar() {

  // create a ref for manipulating the bulma nav-menu node
  const menuEl = useRef(null);

  // toggle the display of the mobile nav-menu 
  const [displayNav, setDisplayNav] = useState(false);
  const navClickHandler = () => {
      if (displayNav === false) {
        menuEl.current.classList.add('is-active');
        setDisplayNav(true);
      } else {
        menuEl.current.classList.remove('is-active');
        setDisplayNav(false);
      }
  };

  return (
    <nav
      className='navbar has-background-white is-fixed-top'
    >
      <div className='navbar-brand'>

        {/* display navbar logo on mobile and desktop */}
        <img 
          className='navbar-logo' 
          src="https://www.rivet.work/wp-content/uploads/2019/10/Rivet_Logo_UnionBlue-300x138.png"
          alt="rivet logo"
        />

        {/* display navbar hamburger menu on mobile only */}
        <div 
          className="navbar-burger"
          aria-label="menu" 
          onClick={() => navClickHandler()}
        >
          <span></span>
          <span></span>
          <span></span>
        </div>

      </div>

      {/* menu displayed on far right in desktop, and hidden dropdown on mobile */}
      <div ref={menuEl} className='navbar-menu'>
        <div className='navbar-end'>
          <div className="navbar-item">
            <Link 
              as="button"
              className='color_shark has-text-weight-bold is-size-5'
              to="/profiles"
              onClick={() => navClickHandler(false)}
            >
              Profile List
            </Link>
          </div>
          <div className="navbar-item">
            <Link 
              as="button"
              className='color_shark has-text-weight-bold is-size-5'
              to="/newprofile"
              onClick={() => navClickHandler(false)}
            >
              Add Employee
            </Link>
          </div>
        </div>
      </div>
    </nav>
  )
}
