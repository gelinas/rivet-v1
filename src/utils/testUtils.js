// testUtils.js rexports @testing-library/react with a customRender method as render
// the customRender method wraps tests in the react-router and redux store providers

import React from 'react';
import { render } from '@testing-library/react'
import { BrowserRouter as Router } from 'react-router-dom';

// redux
import { createStore, applyMiddleware } from 'redux';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import rootReducer from '../reducers';

const store = createStore(rootReducer, applyMiddleware(thunk))

const ProviderWrapper = ({ children }) => {
  return (
    <Provider store={store}>
      <Router>
        {children}
      </Router>
    </Provider>
  )
}

const customRender = (ui, options) =>
  render(ui, { wrapper: ProviderWrapper, ...options })

// re-export everything
export * from '@testing-library/react'

// override render method
export { customRender as render }