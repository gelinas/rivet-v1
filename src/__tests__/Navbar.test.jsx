import React, { useRef } from 'react'

// Testing libraries
import { render, fireEvent } from '../utils/testUtils.js'
import '@testing-library/jest-dom/extend-expect'
import userEvent from '@testing-library/user-event'

// Components to test
import Navbar from '../components/navbar/Navbar'
import { act } from 'react-dom/test-utils'

describe('Tests for Navbar', () => {

  beforeEach(() => {
    // reset mock functions before each test
    jest.clearAllMocks()
  })

  test('should match snapshot', () => {
    const tree = render(
      <Navbar />
    );
    expect(tree).toMatchSnapshot();
  })

  test('should add class is-active to reveal mobile menu when user clicks on burger', () => {
    
    const { getByTestId } = render(<Navbar />);

    // get menu burger icon and navigation menu nodes
    const burgerIcon = getByTestId(/^burger-icon$/i);
    let navMenu = getByTestId(/^nav-menu$/i);

    // test for hidden navigation menu on mobile
    expect(burgerIcon).toBeInTheDocument();
    expect(navMenu).toBeInTheDocument();
    expect(navMenu).not.toHaveClass("is-active");

    // test for navigation menu to be revealed on burgerIcon click
    userEvent.click(burgerIcon);
    navMenu = getByTestId(/^nav-menu$/i);
    expect(navMenu).toHaveClass("is-active");
  })

  test('should remove class is-active to hide mobile menu when user clicks on burger second time', () => {
    
    const { getByTestId } = render(<Navbar />);

    // get menu burger icon and navigation menu nodes
    const burgerIcon = getByTestId(/^burger-icon$/i);
    let navMenu = getByTestId(/^nav-menu$/i);

    // test for hidden navigation menu on mobile
    expect(burgerIcon).toBeInTheDocument();
    expect(navMenu).toBeInTheDocument();
    expect(navMenu).not.toHaveClass("is-active");

    // test for navigation menu to be hidden after two clicks on burgerIcon
    // menu reveal
    userEvent.click(burgerIcon);
    navMenu = getByTestId(/^nav-menu$/i);
    expect(navMenu).toHaveClass("is-active");
    // menu hidden on second click
    userEvent.click(burgerIcon);
    navMenu = getByTestId(/^nav-menu$/i);
    expect(navMenu).not.toHaveClass("is-active");
  })

  test('should remove class is-active to hide mobile menu when user clicks on burger second time', () => {
    
    const { getByTestId } = render(<Navbar />);

    // get menu burger icon and navigation menu nodes
    const burgerIcon = getByTestId(/^burger-icon$/i);
    let navMenu = getByTestId(/^nav-menu$/i);
    let navBrand = getByTestId(/^nav-brand$/i);

    // test for hidden navigation menu on mobile
    expect(burgerIcon).toBeInTheDocument();
    expect(navMenu).toBeInTheDocument();
    expect(navMenu).not.toHaveClass("is-active");

    // test for navigation menu to be hidden after two clicks on burgerIcon
    // menu reveal
    userEvent.click(burgerIcon);
    navMenu = getByTestId(/^nav-menu$/i);
    expect(navMenu).toHaveClass("is-active");
    // menu hidden on outside click
    act(() => userEvent.click(navBrand));
    navMenu = getByTestId(/^nav-menu$/i);
    expect(navMenu).not.toHaveClass("is-active");
  })
})