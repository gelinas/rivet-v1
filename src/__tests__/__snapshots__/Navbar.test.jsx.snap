// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for Navbar should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        class="navbar has-background-white is-fixed-top"
      >
        <div
          class="navbar-brand"
          data-testid="nav-brand"
        >
          <img
            alt="rivet logo"
            class="navbar-logo"
            src="https://www.rivet.work/wp-content/uploads/2019/10/Rivet_Logo_UnionBlue-300x138.png"
          />
          <div
            aria-label="menu"
            class="navbar-burger"
            data-testid="burger-icon"
          >
            <span />
            <span />
            <span />
          </div>
        </div>
        <div
          class="navbar-menu"
          data-testid="nav-menu"
        >
          <div
            class="navbar-end"
          >
            <div
              class="navbar-item"
            >
              <a
                class="color_shark has-text-weight-bold is-size-5"
                href="/profiles"
              >
                <div
                  class="nav_link has-text-right-mobile"
                >
                  Profile List
                </div>
              </a>
            </div>
            <div
              class="navbar-item"
            >
              <a
                class="color_shark has-text-weight-bold is-size-5"
                href="/newprofile"
              >
                <div
                  class="nav_link has-text-right-mobile"
                >
                  Add Employee
                </div>
              </a>
            </div>
          </div>
        </div>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav
      class="navbar has-background-white is-fixed-top"
    >
      <div
        class="navbar-brand"
        data-testid="nav-brand"
      >
        <img
          alt="rivet logo"
          class="navbar-logo"
          src="https://www.rivet.work/wp-content/uploads/2019/10/Rivet_Logo_UnionBlue-300x138.png"
        />
        <div
          aria-label="menu"
          class="navbar-burger"
          data-testid="burger-icon"
        >
          <span />
          <span />
          <span />
        </div>
      </div>
      <div
        class="navbar-menu"
        data-testid="nav-menu"
      >
        <div
          class="navbar-end"
        >
          <div
            class="navbar-item"
          >
            <a
              class="color_shark has-text-weight-bold is-size-5"
              href="/profiles"
            >
              <div
                class="nav_link has-text-right-mobile"
              >
                Profile List
              </div>
            </a>
          </div>
          <div
            class="navbar-item"
          >
            <a
              class="color_shark has-text-weight-bold is-size-5"
              href="/newprofile"
            >
              <div
                class="nav_link has-text-right-mobile"
              >
                Add Employee
              </div>
            </a>
          </div>
        </div>
      </div>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
